cluster-autoscaler/estimator/threshold_based_limiter_test.go

```diff
@@ -0,0 +1,133 @@
+/*
+Copyright 2016 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package estimator
+
+import (
+	"testing"
+	"time"
+
+	apiv1 "k8s.io/api/core/v1"
+
+	"github.com/stretchr/testify/assert"
+)
+
+type limiterRequest interface {
+	apply(*testing.T, EstimationLimiter)
+}
+
+type permissionRequest struct {
+	expectDeny bool
+}
+
+func (p permissionRequest) apply(t *testing.T, l EstimationLimiter) {
+	assert.Equal(t, !p.expectDeny, l.PermissionToAddNode())
+}
+
+type restartRequest struct{}
+
+func (r restartRequest) apply(t *testing.T, l EstimationLimiter) {
+	l.EndEstimation()
+	l.StartEstimation([]*apiv1.Pod{}, nil)
+}
+
+func TestThresholdBasedLimiter(t *testing.T) {
+	testCases := []struct {
+		name            string
+		maxNodes        int
+		maxDuration     time.Duration
+		startDelta      time.Duration
+		operations      []limiterRequest
+		expectNodeCount int
+	}{
+		{
+			name:     "no limiting happens",
+			maxNodes: 20,
+			operations: []limiterRequest{
+				permissionRequest{false},
```

Comment:
nit: I think instead of using the pattern of applying operations here it would be more readable to just have a list of `func()` here that are invoked, just so that they have meaningful names in each test case like so:

```
operations: []operation{
  expectDeny,
  restartLimiter,
  expectAllow,
}
```

Reply:
You're right, that's much better. Updated the test.